cabal-install/src/Distribution/Client/VCS.hs

```diff
@@ -608,11 +612,33 @@ vcsGit =
               { progInvokeCwd = Just cwd
               }
 
+        -- Beware: if the user supplied revision for the source repository
+        -- package is /not/ a full hash, then we cannot fetch it, which means
+        -- we cannot do a shallow clone (--depth=1).
+        -- See #10605.
+        doShallow
+          | Nothing <- srpTag =
+              -- No tag, OK for shallow
+              True
+          -- full hashes are 40 characters
+          | Just tg <- srpTag
+          , length tg >= 40 =
```

Comment:
It should be _exactly_ 40 characters and all hexadecimal characters; e.g. `48c42600594a381af18afb24d8eb857396876f0ff` is not valid and `48c42600594a38Iaf18afb24d8eb857396876f0f` is not valid.